libcrmcommon, tools: Various refactors and improvements

nrwahl2 opened the pull request:
In preparation for #2847 to use formatted output in crm_error and move to libpacemaker

nrwahl2 commented on include/crm/common/results.h:
```diff
@@ -325,6 +325,29 @@ enum pcmk_exec_status {
+ * specifies how to interpret the integer.
+ */
+typedef enum pcmk_result_type_e {
+    pcmk_result_legacy       = 0,   //!< Legacy API function return code
```
Do we need to add any kind of deprecation warning here? I didn't, because this enum value per se isn't deprecated; it signifies that some integer should be interpreted as a deprecated type.

clumens replied:
No, I don't think that this needs a warning.

Note that unless you actually care about the values, you don't need to assign values to the things in an enum.  The C standard guarantees the first element is 0 and all subsequent elements are one more than the previous (section 6.7.2.2 in C17, if you really care).

nrwahl2 replied:
Thanks! I did know that the standard guarantees that. I didn't know if we had any preference to be explicit; we have a few enums that explicitly assign either the first element or all elements, even though they're just the sequence of consecutive integers starting with 0.

I see there are also a bunch that don't explicitly assign values to members, so I'll drop the assignments here.

kgaillot replied:
A good reason to keep the explicit assignments in this particular case is that we'll want to deprecate and drop pcmk_result_legacy after the legacy rc's are dropped, and we'll need the explicit assignments for the remaining values to avoid them changing at that time

Another situation where explicit assignments are useful is when the values are ordered and we might want to insert values in the middle someday (in which case the initial values should be something like 10, 20, 30, ...).

nrwahl2 replied:
Ah good point. I'll take care of that

nrwahl2 commented on lib/pacemaker/pcmk_result_code.c:
```diff
@@ -0,0 +1,215 @@
+                 (show_desc? desc : NULL));
+    out->quiet = quiet_orig;
+
+    return rc;
```
I'm on the fence about whether we should return an error code like `ENXIO` (mapped to `CRM_EX_NOSUCH`) if we our lookup returns an unknown error. Currently, we're returning `pcmk_rc_ok` because the query didn't do anything off-the-wall like returning a null pointer -- it printed the string "Unknown" or something like that. Further, `crm_error` doesn't return an error in this case.

But it seems like an arbitrary caller might want to check for an error code to know whether we found a real value for the code we're querying, or whether we simply printed "Unknown".

`crm_error` can always ignore an error code from `pcmk__show_return_code()`. However, `crm_error` might want to do something in a hypothetical case where `get_strings()` gives us a null pointer, even if it ignores the result "Unknown". Could use two different error codes but `ENXIO` seems like the closest match for both. May have to make a new one.

nrwahl2 replied:
`crm_error` itself probably should return an error in case of "Unknown", but changing the exit codes could be a regression.

nrwahl2 commented on include/crm/common/output_internal.h:
```diff
@@ -28,7 +28,7 @@ extern "C" {
 
 
-#  define PCMK__API_VERSION "2.22"
+#  define PCMK__API_VERSION "2.24"
```
This is so that when #2836 gets merged we won't forget to bump for this one too.

nrwahl2 commented on tools/crm_error.c:
```diff
@@ -140,8 +65,8 @@ main(int argc, char **argv)
-    const char *name = NULL;
-    const char *desc = NULL;
+
+    pcmk_result_type_t type = pcmk_result_legacy;
```
This is the current default. We need to figure out when and how we want to change it without introducing an unplanned regression for anything external that might depend on `crm_error`.

nrwahl2 commented:
retest this please

nrwahl2 commented on tools/crm_error.c:
```diff
@@ -86,6 +86,14 @@ main(int argc, char **argv)
         goto done;
     }
 
+    if (options.as_exit_code && options.as_rc) {
```
The current lack of a check is a minor bug and no users should be specifying both options together -- but https://projects.clusterlabs.org/T261 makes me wonder if we need to wait on this.

nrwahl2 replied:
I'd like to think no one is relying on both of them combined being treated as `-X`. If they are, I would say it's a bug in **their** code...

And right, I think with text output, they're more likely to be checking the output than the return code anyway, so I don't know how much better an extra warning line in the output would be for them, compared to erroring and bailing out.

I don't have strong feelings and can change this to a warning if we want to.

kgaillot replied:
Yeah we get bit a lot when we assume no one's relying on something :)

Also, I'm not sure we want to reject this. A lot of OS commands accept contradictory arguments with the last one winning. Here that would require redoing the option values as an enum, but that's already in here later ...

nrwahl2 replied:
Last one winning would be fine and I can switch to that. The current behavior is not order-dependent -- if `-X` is present, we treat the value as an exit code no matter what: `-X`, `-rX`, `-Xr`. It's not intuitive that `-X` should automatically win.

kgaillot replied:
Yeah last one wins makes sense

clumens commented on tools/crm_error.c:
```diff
@@ -137,37 +225,45 @@ main(int argc, char **argv)
+        exit_code = pcmk_rc2exitc(rc);
+    }
+
     pcmk__output_and_clear_error(error, NULL);
```
You should pass `out` here as the second argument to pcmk__output_and_clear_error.

nrwahl2 replied:
Oops

clumens commented on include/pacemaker.h:
```diff
@@ -186,6 +186,41 @@ int pcmk_list_nodes(xmlNodePtr *xml, char *node_types);
+int pcmk_list_result_codes(xmlNodePtr *xml, pcmk_result_type_t type,
+                           bool show_name, bool show_desc);
+
 #ifdef BUILD_PUBLIC_LIBPACEMAKER
```
I'm going to leave it up to @kgaillot to decide whether these things should be above or below BUILD_PUBLIC_LIBPACEMAKER.

I find it a little confusing that we have both `pcmk_` functions and that macro guard, when anything that's `pcmk_` is already public API.  That fact means these new functions would still be public, which means changing to a set of flags later would be difficult.  So it's probably worth starting with flags.

nrwahl2 replied:
Ah fair enough. I turned them into flags; they'll go in the next update.

Agreed about `BUILD_PUBLIC_LIBPACEMAKER`. I think commit 27677a6d drew it away from its intended purpose.

kgaillot replied:
The stonith_admin API was the first one to be added to libpacemaker, and it wasn't ready for release (still isn't), so BUILD_PUBLIC_LIBPACEMAKER was added to prevent it from going public just yet, even though it is intended as public API eventually.

Then we added other APIs that were final, so they didn't need to be guarded, and the name became misleading. Something like PCMK__ENABLE_FENCE_API would be better now.

FYI the main thing holding up the fencing API is that we don't want to drag stonith_t into libpacemaker. Eventually the fencer will be converted to use pcmk_ipc_api_t instead, and then we can decide what the public API should look like. Another possibility would be to just drop the stonith_t arguments and have each public API create its own connection.

clumens commented on include/pcmki/pcmki_result_code.h:
```diff
@@ -0,0 +1,13 @@
+#ifndef PCMK__PCMKI_PCMKI_RESULT_CODE__H
```
Remember the copyright boilerplate at the top of the file.

nrwahl2 replied:
Done, thanks. Like I said in #2836, the existing one I used as a template didn't have the copyright. Fixed that in a new commit in #2836.

clumens commented on lib/pacemaker/pcmk_result_code.c:
```diff
@@ -0,0 +1,215 @@
+            const char **desc)
+{
+    CRM_ASSERT((type == pcmk_result_legacy) || (type == pcmk_result_pcmk_rc)
+               || (type == pcmk_result_crm_exitcode));
```
It can sometimes be confusing, but you could also write this like:
```
CRM_ASSERT(type <= pcmk_result_crm_exitcode);
```

nrwahl2 replied:
I probably will, just because there's a precedent. It's already used quite a lot within Pacemaker.

To be honest, that convention always gives me a headache, because understanding it requires going back and checking the order of the enum. An enum is technically ordered but it's usually fairly arbitrary, like here -- `pcmk_result_ocf_exitcode` is not greater than `pcmk_result_crm_exitcode` in any semantically meaningful way.

nrwahl2 replied:
Didn't see this. I'll revert and update.

nrwahl2 commented on tools/crm_error.c:
```diff
@@ -120,24 +120,31 @@ main(int argc, char **argv)
             }
-            get_strings(rc, &name, &desc);
+            get_strings(current, type, &name, &desc);
+
```
Good catch. It changes in the move from crm_error to libpacemaker. This is where I accidentally backported it to crm_error while moving the for-to-while stuff into crm_error (after writing it in libpacemaker).

nrwahl2 commented:
Before the next push, I'll split the new public API functions into a separate commit, like @kgaillot said in #2836 for rule checks.

nrwahl2 commented on tools/crm_error.c:
```diff
@@ -56,12 +56,88 @@ get_strings(int rc, const char **name, const char **str)
+    if (desc != NULL) {
+        crm_xml_add(node, "description", desc);
+    }
+
```
Oh that's a neat helper.

nrwahl2 commented:
Summary of updates to address review and some other things I noticed along the way:
* Fixed call to get_strings() with four arguments before move to libpacemaker (https://github.com/ClusterLabs/pacemaker/pull/2838#discussion_r957689645). It didn't affect the final result, but one of the intermediate commits would have failed to compile.
* Removed explicit values from the `pcmk_result_type_t` enum definition
* Changed asserts to `<=` for `pcmk_result_type_t`. I compromised on my readability concern by commenting the meaning above the assert. (https://github.com/ClusterLabs/pacemaker/pull/2838#discussion_r957746066)
* Split new public API functions into separate commits
* Added all XML string attributes using `pcmk__output_create_xml_node()` (https://github.com/ClusterLabs/pacemaker/pull/2838#discussion_r957695205). (I wish the int could come first while still using this handy function, but it doesn't matter at all.)
* Added copyright to pcmki header, and updated in `crm_error.c`.
* Fixed `pcmk__output_and_clear_error()` call by passing `out` instead of `NULL`.
* Added `enum pcmk_rc_disp_flags` and updated the API functions to use them instead of two booleans. This cleaned up the internal API a little bit and made it more flexible. I'm still using the `out->quiet` flag as a hack to control whether the output looks like the `crm_error -- <rc>` output format (`[<name> - ]<desc>`) or the `crm_error --list` output format (`<code>: [<name> <spaces>] [<desc>]`).

CI has been extremely slow lately (IIRC mostly bottlenecked on RHEL 8.6 ppc64le). This passes cts-cli and `make rpm` succeeds on my machine, so fingers crossed.

clumens commented:
>     * Added all XML string attributes using `pcmk__output_create_xml_node()` ([Use formatted output in crm_error and move functionality to libpacemaker #2838 (comment)](https://github.com/ClusterLabs/pacemaker/pull/2838#discussion_r957695205)). (I wish the int could come first while still using this handy function, but it doesn't matter at all.)

It would be possible to do this, but it would require some sort of type code in the argument list.  It would look something like this:
```
pcmk__output_create_xml_node(out, "whatever",
                             "attr1", pcmk__ty_str, "value1",
                             "attr2", pcmk__ty_int, 47,
                             NULL);
```

And then in pcmk__xe_set_propv we'd just use that type code to determine what to put into the call to va_arg.  It wouldn't be hard, but maybe the callers end up more verbose for little payoff.  @kgaillot any thoughts?

kgaillot commented on tools/crm_error.c:
```diff
@@ -124,9 +122,10 @@ main(int argc, char **argv)
             } else if(options.with_name) {
-                printf("% .*d: %-26s  %s\n", width, rc, name, desc);
+                // Room for four digits and a negative sign, space-padded
+                printf("% 5d: %-30s  %s\n", rc, name, desc);
```
Rather than expand the fields, I think we should loop once through all the values and find the longest name and value, and use the relevant widths.

nrwahl2 replied:
This will be easier if we wait for until the end of the commits in this PR, rather than doing it in `crm_error.c` early on. Later, I add an internal constant `pcmk__n_rc` that exposes the number of negative return codes outside of `results.c`. It replaces `PCMK__N_RC`.

I think if we're going to get name and value widths dynamically, it only makes sense to base it on the standard Pacemaker return codes. In practice, the widest name and value of that class should always be wider than the widest of any other class. I'll do this in the next update.

The alternative is to add one or more "width" arguments to the formatter functions and let the caller loop over the relevant values to determine the width(s) it needs. I don't like that, because (1) those arguments would be ignored by the XML format function., and (2) less importantly, it would add one or two extra params to deal with.

nrwahl2 replied:
I just moved all the refactor and non-API feature stuff to the beginning of the commit chain and did them in `crm_error.c` instead of in `pcmk_result_code.c` and `pcmk_output.c`, for easier review.

kgaillot commented on tools/crm_error.c:
```diff
@@ -102,17 +102,15 @@ main(int argc, char **argv)
         // 256 is a hacky magic number that "should" be enough
         if (options.as_rc) {
             start = pcmk_rc_error - 256;
             end = PCMK_CUSTOM_OFFSET;
```
If we're going to try to improve this, we might as well go further and choose start more reliably. We could define a new alias enum value for the lowest-numbered value (with a loud comment reminding us to update it), or you could loop until you hit the default message.

end is probably impossible to determine reliably since it's the largest system errno, and there's no symbol defined for that and no guarantee they're consecutive. I'm not sure if there's a better approach than this.

nrwahl2 replied:
I did that in the (currently) second-to-last commit in this PR. Title in case the commit itself gets overwritten: "Refactor: libpacemaker: Improve loop bounds in pcmk__list_result_codes()"

nrwahl2 replied:
I just moved all the refactor and non-API feature stuff to the beginning of the commit chain and did them in `crm_error.c` instead of in `pcmk_result_code.c` and `pcmk_output.c`, for easier review.

kgaillot replied:
Let's split this into two PRs at "Use formatted output", this one has gotten difficult to follow

nrwahl2 replied:
I misread you there and moved the stuff *after* formatted output. I'll move the formatted output commits over there shortly.

kgaillot commented on tools/crm_error.c:
```diff
@@ -120,24 +120,31 @@ main(int argc, char **argv)
+                 * so skip them
+                 */
+                current = pcmk_rc_ok;
+                continue;
```
A for loop with some extra bumps inside is actually not that strange, e.g. you see it a lot with command-line argument processing when an argument takes an value. But I'm fine with redoing this one.

kgaillot commented on tools/crm_error.c:
```diff
@@ -56,12 +56,83 @@ get_strings(int rc, const char **name, const char **str)
+        } else if (name != NULL) {
+            pcmk__formatted_printf(out, "%s\n", name);
+        } else if (desc != NULL) {
+            pcmk__formatted_printf(out, "%s\n", desc);
```
Why not use out->info()?

nrwahl2 replied:
This was probably a remnant of a different approach. `out->info()` would be suitable.

nrwahl2 replied:
Fixed

nrwahl2 replied:
Now I remember. It's because we tell the formatter to use this compact format (instead of the columnar format that we use for lists) by setting `out->quiet = true`.

When `out->quiet` is true for a text output object, `out->info()` is a no-op. `pcmk__formatted_printf()` forces it to print the output. There's a good bit of precedent for that in `pcmk_output.c` already.

kgaillot replied:
That's risky since this is used as the default implementation, not just the text implementation. (What happens with --output-as=none?)

Not sure what the best alternative is though

nrwahl2 replied:
Good catch. That's a bug between the way `pcmk__formatted_printf()` is written and the way it's used... it's defined in `output_text.c` but it's used in several default formatter functions. So it can also get used with output type none.

Added a patch in #2847 
https://github.com/ClusterLabs/pacemaker/pull/2847/commits/b4768fa7db2df181328babcea89a61c66d19dfe5#r961161871

@clumens FYI

kgaillot reviewed (commented):
Not a comprehensive review, but a first look ...

nrwahl2 commented on tools/crm_error.c:
```diff
@@ -136,6 +136,10 @@ main(int argc, char **argv)
+            if (pcmk__strcmp(processed_args[lpc], "--", pcmk__str_none) == 0) {
+                continue;
+            }
+
```
Yes, it's a regression introduced by that commit. Just verified. I'll update the commit message to mention it.

Before, it works just as you'd expect. After, it does this:
```
$ crm_error -n -- 1
EPERM - Operation not permitted
$ crm_error -n -- -1
EPERM - Operation not permitted
EPERM - Operation not permitted
$ crm_error -n -- -1010
EPERM - Operation not permitted
Unknown - Unknown error 1010
```

Not exactly the same as the current behavior (below) but close enough:
```
# crm_error -n -- -1010
Unknown - Error
Unknown - Unknown error 1010
```

nrwahl2 commented:
The latest update is mostly reordering commits so that the refactors take place in `crm_error` early on. I also dropped the "reject `-X` with `-r`" commit and replaced it with a commit to use an enum and callback. The last option for result type wins.

kgaillot commented:
> ```
> pcmk__output_create_xml_node(out, "whatever",
>                              "attr1", pcmk__ty_str, "value1",
>                              "attr2", pcmk__ty_int, 47,
>                              NULL);
> ```
> 
> And then in pcmk__xe_set_propv we'd just use that type code to determine what to put into the call to va_arg. It wouldn't be hard, but maybe the callers end up more verbose for little payoff. @kgaillot any thoughts?

I think it would make the calls too long and confusing

kgaillot commented on tools/crm_error.c:
```diff
@@ -132,13 +133,21 @@ main(int argc, char **argv)
+                end = CRM_EX_MAX + 1;
+                break;
+            default:    // Make the compiler happy
+                break;
```
I wonder if this is worth abstracting too e.g. pcmk_result_bounds(result_kind, &start, &end)

kgaillot reviewed (commented):
This should do it

kgaillot commented on include/crm/common/results.h:
```diff
@@ -326,6 +326,29 @@ enum pcmk_exec_status {
+    pcmk_result_ocf_exitcode  = 3,  //!< Exit status code for resource agents
+    pcmk_result_exec_status   = 4,  //!< Execution status
+    pcmk_result_bz2           = 5,  //!< Return code from bzlib
+} pcmk_result_type_t;
```
@clumens , do you have any thoughts on a new enum vs reusing GQuark for this?

clumens replied:
That's an interesting idea and I think I like it.  GQuarks are made for identifying error domains, which is essentially what we are doing here.  Off the top of my head, I can't think of any reasons not to.

kgaillot replied:
I think the biggest downside is how to compare the value. I don't think we can use switch or `==`; we'd have to use `g_intern_string()` or something like that.

It also means the relevant public APIs will require glib. We already use GList and GHashTable elsewhere, so it's not that big of a deal, but I do like to minimize how much glib we expose.

clumens replied:
You should just be able to compare GQuarks.  They're just typedef'd to a guint32 (at least in this one version of the documentation I'm seeing - can't imagine that's changed).

nrwahl2 replied:
The doc says GQuarks are meant to be unique per string, so I think that would work. Let me know if we want to go that route. I guess it would require making the two existing GQuarks public (in name and in header) and then adding a couple of new ones.

The only thing I have in favor of an enum is that the meaning is especially clear (doxygen comment would solve that) and the values are constrained to what's defined.

kgaillot replied:
Another consideration is that the quark constants are actually function calls, so that's a good bit of overhead vs enums. Since we're only planning on using this in crm_error, I think that's OK. Hopefully it won't bite us later.

Let's just add one new one for legacy codes since that's all crm_error will need right now. It would be easy enough to add others later.

nrwahl2 replied:
We can get around this with `if` statements, but this is annoying.
```
results.c:51:9: error: case label does not reduce to an integer constant
   51 |         case PCMK_LEGACY_ERROR:
      |         ^~~~
results.c:55:9: error: case label does not reduce to an integer constant
   55 |         case PCMK_RC_ERROR:
      |         ^~~~
results.c:59:9: error: case label does not reduce to an integer constant
   59 |         case PCMK_EXITC_ERROR:
```

nrwahl2 replied:
It's also impossible to use these as initializers.
```
struct {    
    gboolean with_name;
    gboolean do_list;
    GQuark result_type; // How to interpret result codes
} options = {   
    .result_type = PCMK_LEGACY_ERROR,
}; 

In file included from ../include/crm/services.h:23,
                 from ../include/crm/lrmd.h:16,
                 from ../include/crm_internal.h:37,
                 from crm_error.c:10:
../include/crm/common/results.h:81:29: error: initializer element is not constant
   81 | #define PCMK_LEGACY_ERROR   pcmk_legacy_error_quark()
      |                             ^~~~~~~~~~~~~~~~~~~~~~~
crm_error.c:26:20: note: in expansion of macro 'PCMK_LEGACY_ERROR'
   26 |     .result_type = PCMK_LEGACY_ERROR,
      |                    ^~~~~~~~~~~~~~~~~
../include/crm/common/results.h:81:29: note: (near initialization for 'options.result_type')
   81 | #define PCMK_LEGACY_ERROR   pcmk_legacy_error_quark()
      |                             ^~~~~~~~~~~~~~~~~~~~~~~
crm_error.c:26:20: note: in expansion of macro 'PCMK_LEGACY_ERROR'
   26 |     .result_type = PCMK_LEGACY_ERROR,
      |                    ^~~~~~~~~~~~~~~~~
```

I'm almost finished (I think), but I'm going to hold off in case we want to rethink using the quarks for this. So far, it seems like it's 100% possible to use the quarks to identify the result code families, but:
1. It will make the code clunkier (can't use it in `switch` or static initializers), and
2. It requires function call overhead (probably not a big deal but it was noted above).

kgaillot replied:
Yeah, let's stick with enum

kgaillot replied:
I forgot to mention, enums don't need _e at the end or typedef, though you'll see both in legacy code. "enum X" is sufficient and ideal to indicate X is an enum.

nrwahl2 commented on tools/crm_error.c:
```diff
@@ -120,17 +120,33 @@ main(int argc, char **argv)
+         */
+        char *longest_rc_str = pcmk__itoa(pcmk_rc_error - pcmk__n_rc + 1);
+        int code_width = strlen(longest_rc_str);
+        int name_width = 0;
```
Oh that is neat.

nrwahl2 replied:
> Should use size_t for all these

They have to be `int`s for the width specifier. The return values are `size_t` but we're not using them for anything other than width specifiers. So I think it makes more sense to use an `int` variable than to use a `size_t` and cast back to `int` in the `printf()` call.

```
pcmk_output.c:1963:34: error: field width specifier ‘*’ expects argument of type ‘int’, but argument 3 has type ‘size_t’ {aka ‘long unsigned int’} [-Werror=format=]
 1963 |         return out->info(out, "% *d: %-*s  %s", code_width, code, name_width, name,
      |                                ~~^~             ~~~~~~~~~~
      |                                  |              |
      |                                  int            size_t {aka long unsigned int}
```

kgaillot replied:
Ah right. Static analysis is the main reason to care, since we know it's not a problem here, so I don't mind leaving them as ints. But in case you're interested there are two portable approaches:

* use size_t with %lld and cast it to long long.
* use size_t for initial values, error if they're out of int range, and cast them to int otherwise.

nrwahl2 commented:
Updates:
* Drop second `pcmk` from `pcmk_result_pcmk_rc`
* Combine `pcmk_result_crm_exitcode` and `pcmk_result_ocf_exitcode` into `pcmk_result_exitcode`
* Drop `pcmk_result_bz2` from the enum
* Use size_t for `pcmk__n_rc`
* Use the `snprintf()` trick to get number's string width, and cast explicitly to int
* Functionized getting upper and lower bounds of a result code family (`pcmk__result_bounds()`)

I worked on the change to quarks, but I hit a couple of hiccups that caused me to reconsider. I saved those changes in a separate branch if we still want to use quarks as result types.
* https://github.com/ClusterLabs/pacemaker/pull/2838#discussion_r964491049

kgaillot commented on tools/crm_error.c:
```diff
@@ -44,19 +60,24 @@ static GOptionEntry entries[] = {
+        default:
+            break;
     }
 }
```
I think this function is now worth moving to libcrmcommon as public API

kgaillot commented on lib/common/results.c:
```diff
@@ -25,6 +25,48 @@
+ *       system.
+ */
+int
+pcmk__result_bounds(pcmk_result_type_t type, int *lower, int *upper)
```
Internal API is good since these are not known precisely

kgaillot reviewed (commented):
Last minor comments

nrwahl2 commented:
Updates:
* Removed trailing `_e` from enum name.
* Moved `get_strings()` to public API as `pcmk_result_get_strings()`, accepting a type argument.
* Inserted commit "Remove main-scope rc variable in crm_error" to remove an unnecessary "rc" variable that looked like a function return code but wasn't acting like one. Some later commits required variable name changes after that.

kgaillot commented on include/crm/common/results.h:
```diff
@@ -326,6 +326,27 @@ enum pcmk_exec_status {
+ * These values are useful for specifying how an integer result code should be
+ * interpreted in situations involving a generic integer value. For example, a
+ * function that can process multiple types of result codes might accept an
+ * arbitrary integer argument along with a \c pcmk_result_type_e argument that
```
whoops :)

kgaillot commented on lib/common/results.c:
```diff
@@ -25,6 +25,47 @@
+pcmk_result_get_strings(int code, enum pcmk_result_type type, const char **name,
+                        const char **desc)
+{
+    CRM_ASSERT((name != NULL) && (desc != NULL));
```
For flexibility, we should allow name and desc to be NULL, and set them only if not (in case a caller needs just one of them)
